File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.cc
#include<iostream>
#include<queue>
#include<string>
using namespace std;

int row, col, minsikY, minsikX, ans;
int arr[51][51];
bool visit[51][51][65];
int dy[4] = { 0, 0, -1, 1 }, dx[4] = { -1, 1, 0, 0 };

typedef pair<int, int> pii;

void bfs() {
	queue<pair<pii, pii>> q;
	q.push({ {minsikY, minsikX}, {0, 0} }); // y, x, 이동 횟수, 키
	visit[minsikY][minsikX][0] = true;

	while (!q.empty()) {
		int currY = q.front().first.first;
		int currX = q.front().first.second;
		int currDist = q.front().second.first;
		int currKeys = q.front().second.second;
		q.pop();

		if (arr[currY][currX] == 1) {
			ans = currDist;
			return;
		}

		for (int i = 0; i < 4; i++) {
			int ny = currY + dy[i];
			int nx = currX + dx[i];
			int newKeys = currKeys;

			if (ny < 0 || nx < 0 || ny >= row || nx >= col) continue;
			if (arr[ny][nx] == -1 || visit[ny][nx][newKeys]) continue;

			//cout << currKeys << " " << arr[ny][nx] << " " << arr[ny][nx] - 'A' << " " << (currKeys & (1 < (arr[ny][nx] - 'A'))) << "\n";
			// A~F 칸인데 키가 없는 경우
			if (arr[ny][nx] >= 'A' && arr[ny][nx] <= 'F'
				&& (currKeys & (1 << (arr[ny][nx] - 'A'))) == 0) continue;

			if (arr[ny][nx] >= 'a' && arr[ny][nx] <= 'f') {
				newKeys = (currKeys | (1 << (arr[ny][nx] - 'a')));
			}

			q.push({ {ny, nx}, {currDist + 1, newKeys} });
			visit[ny][nx][newKeys] = true;
		}
	}
}


int main() {
	ios::sync_with_stdio(0);
	cin.tie(nullptr);
	cout.tie(nullptr);

	cin >> row >> col;
	
	/*cout << 'A' - 0 << "\n";
	cout << 'A' - 0 - 'A'<< "\n";
	cout << 'F' - 0 - 'A' << "\n";
	return 0;*/

	for (int i = 0; i < row; i++) {
		string str;
		cin >> str;
		for (int j = 0; j < col; j++) {
			if (str[j] == '.') continue;
			if (str[j] == '#') {
				arr[i][j] = -1;
				continue;
			}

			if (str[j] == '0') {
				minsikY = i;
				minsikX = j;
				continue;
			}

			if (str[j] == '1') {
				arr[i][j] = 1;
				continue;
			}

			// 알파벳들
			arr[i][j] = str[j] - 0;
		}
	}

	bfs();

	if (ans == 0) cout << "-1\n";
	else cout << ans << "\n";

	return 0;
}

File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.java

import java.util.*;
import java.io.*;

public class Main {

	static int row, col, minsikY, minsikX, ans;
	static int arr[][];
	static int dy[] = {0, 0, -1, 1}, dx[] = {-1, 1, 0, 0};
	static boolean visit[][][];
	
	public static void main(String[] args) throws Exception{
		
//		System.out.println('A' - 0);
//		System.out.println('a' - 0);
//		System.out.println('a' - 'a');
//		System.out.println('A' - 'a');
//		System.out.println('a' - 'A');
//		System.out.println(1 << 'a' - 'a');
//		System.out.println(1 << 'b' - 'a');
//		System.out.println(1 << 'c' - 'a');
//		System.out.println(1 << 'd' - 'a');
//		System.out.println(1 << 'e' - 'a');
//		System.out.println(1 << 'f' - 'a');
		
//		System.out.println(2 & (1 << 'F' - 'a'));
//		System.out.println(1 << 'F' - 'a');
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		row = Integer.parseInt(st.nextToken());
		col = Integer.parseInt(st.nextToken());
		
		arr = new int[row + 1][col + 1];
		visit = new boolean[row + 1][col + 1][200];
		
		for(int i = 0; i< row; i++) {
			st = new StringTokenizer(br.readLine());
			String str = st.nextToken();
			for(int j = 0; j< col; j++) {
				char c = str.charAt(j);
				if(c == '.') continue;
				if(c == '#') {
					arr[i][j] = -1;
					continue;
				}
				if(c == '0') {
					minsikY = i;
					minsikX = j;
					continue;
				}
				if(c == '1') {
					arr[i][j] = 1;
					continue;
				}
				arr[i][j] = c - 0;
			}
		}
		
		bfs();
		
		System.out.println(ans == 0 ? -1 : ans);

	}
	
	static void bfs() {
		Queue<int[]> q = new ArrayDeque<>();
		visit[minsikY][minsikX][0] = true;
		q.offer(new int[] {minsikY, minsikX, 0, 0}); // y, x, 거리, 키
		
		while(!q.isEmpty()) {
			int currY = q.peek()[0];
			int currX = q.peek()[1];
			int currDist = q.peek()[2];
			int keys = q.poll()[3];
			
//			System.out.println("(" + currY + ", " + currX + ") " + currDist + " " + keys);
//			for(int i = 0; i< row; i++) {
//				for(int j = 0; j< col; j++) {
//					if(i == currY && j == currX) {
//						System.out.print("X ");
//						continue;
//					}
//					System.out.print(arr[i][j] + " ");
//				}
//				System.out.println();
//			}
//			System.out.println();
			
			if(arr[currY][currX] == 1) {
				ans = currDist;
				return;
			}
			
			for(int i = 0; i< 4; i++) {
				int ny = currY + dy[i];
				int nx = currX + dx[i];
				int newKey = keys;
				
				if(ny < 0 || nx < 0 || ny >= row || nx >= col) continue;
				//System.out.println(keys + " " + visit[ny][nx] + " "+(visit[ny][nx] & keys));
				if(arr[ny][nx] == -1 || (keys != 0 && visit[ny][nx][keys]) || (keys == 0 && visit[ny][nx][0] == true)) continue; // 벽
				if(arr[ny][nx] >= 'A' && arr[ny][nx] <= 'F' && ((keys & (1 << arr[ny][nx] - 'a')) == 0)) continue; // 키가 없을때
				
				// 키 획득
				if(arr[ny][nx] >= 'a' && arr[ny][nx] <= 'f') newKey =  keys | (1 << (arr[ny][nx] - 'a'));  
					//hasKey[arr[ny][nx] - 'a'] = true;
				
				q.offer(new int[] {ny, nx, currDist + 1, newKey});
				if(newKey != 0) visit[ny][nx][newKey] = true;
				else visit[ny][nx][0] = true;
				
			}
		}
		
	}

}












